drivers/nrf24l01p/nrf24l01p_communication.c

```diff
@@ -0,0 +1,217 @@
+#include "xtimer.h"
+#include "nrf24l01p_constants.h"
+#include "nrf24l01p_communication.h"
+
+#define ENABLE_DEBUG    (0)
+
+#ifndef MIN
+#define MIN(a, b) (((a) < (b)) ? (a) : (b))
+#endif
+
+#define SPI_BUS     (dev->params.spi)
+#define SPI_PIN_CS  (dev->params.pin_cs)
+#define DUMMY       NULL
+
+#define NRF24L01P_REG_MAX_WIDTH NRF24L01P_MAX_ADDR_WIDTH
+
+static void swap_bytes_inplace(void *buffer, size_t len)
+{
+    uint8_t *bytes = (uint8_t *)buffer;
+
+    for (size_t i = 0; i < len / 2; i++) {
+        uint8_t tmp = bytes[i];
+        bytes[i] = bytes[len - 1 - i];
+        bytes[len - 1 - i] = tmp;
+    }
+}
+
+int nrf24l01p_acquire(nrf24l01p_t *dev)
+{
+    DEBUG("[nrf24l01p] acquire ...\n");
+    int a = spi_acquire(dev->params.spi, dev->params.pin_cs, SPI_MODE_0,
+                        dev->params.spi_clk);
+    #ifndef NDEBUG
+    dev->have_spi_access = 1;
+    #endif
+    DEBUG("[nrf24l01p] acquired!\n");
+    return a;
+}
+
+void nrf24l01p_release(nrf24l01p_t *dev)
+{
+    spi_release(dev->params.spi);
+    #ifndef NDEBUG
+    dev->have_spi_access = 0;
+    #endif
+    DEBUG("[nrf24l01p] released\n");
+}
+
+uint8_t nrf24l01p_read_reg(nrf24l01p_t *dev, uint8_t reg, uint8_t *dest,
+                           size_t len)
+{
+    uint8_t cmd = NRF24L01P_CMD__R_REGISTER(reg);
+    uint8_t status = spi_transfer_byte(SPI_BUS, SPI_PIN_CS, true, cmd);
+
+    spi_transfer_bytes(SPI_BUS, SPI_PIN_CS, false, DUMMY, dest, len);
+    if (len > 1) {
+        swap_bytes_inplace(dest, len);
+    }
+    return status;
+}
+
+uint8_t nrf24l01p_write_reg(nrf24l01p_t *dev, uint8_t reg, const uint8_t *src,
+                            size_t len)
+{
+    uint8_t cmd = NRF24L01P_CMD__W_REGISTER(reg);
+    uint8_t status;
+
+    if (len > 1) {
+        len = MIN(len, NRF24L01P_REG_MAX_WIDTH);
+        uint8_t lsrc[len];
+        memcpy(lsrc, src, len);
+        swap_bytes_inplace(lsrc, len);
+        status = spi_transfer_byte(SPI_BUS, SPI_PIN_CS, true, cmd);
+        spi_transfer_bytes(SPI_BUS, SPI_PIN_CS, false, lsrc, DUMMY, len);
+    }
+    else {
+        status = spi_transfer_byte(SPI_BUS, SPI_PIN_CS, true, cmd);
+        spi_transfer_bytes(SPI_BUS, SPI_PIN_CS, false, src, DUMMY, len);
+    }
+    return status;
+}
+
+uint8_t nrf24l01p_read_rx_payload(nrf24l01p_t *dev, void *dest, size_t len)
+{
+    uint8_t cmd = NRF24L01P_CMD__R_RX_PAYLOAD;
+    uint8_t status = spi_transfer_byte(SPI_BUS, SPI_PIN_CS, true, cmd);
+
+    spi_transfer_bytes(SPI_BUS, SPI_PIN_CS, false, &cmd, dest, len);
+    if (len > 1) {
+        swap_bytes_inplace(dest, len);
+    }
+    return status;
+}
+
+uint8_t nrf24l01p_write_tx_payload(nrf24l01p_t *dev, const void *src,
+                                   size_t len)
+{
+    uint8_t cmd = NRF24L01P_CMD__W_TX_PAYLOAD;
+    uint8_t status;
+
+    if (len > 1) {
+        len = MIN(len, NRF24L01P_MAX_PAYLOAD_WIDTH);
+        uint8_t lsrc[len];
+        memcpy(lsrc, src, len);
+        swap_bytes_inplace(lsrc, len);
```

Comment:
I don't really understand what the purpose of this is. The byte order is again changed in `nrf24l01p_read_rx_payload()`. So wouldn't it have the same result to just drop the `swap_bytes_inpalce()` in both functions?

Reply:
The `nrf24l01p` expects that the LSByte of payload or register values is written first and
when you read the value you´ll get the LSByte first as well. I don´t know why they designed it that way, since it just wastes CPU cycles in my eyes.

So if you want the paylaod `abc` to be transmitted, you must write `cba` to the SPI.
And if you have received `def` and you read it via SPI, it will output `fed`. 

I am not sure what the result will be if I read my own tx payload that i have just written.
However, I think there is no reason to do that.

I don´t think a can drop that.